File: problem/Delete_Columns_Sorted_II/minDeletionSize.c
int minDeletionSize(char **strs, int strsSize)
{
    if (strsSize <= 1)
        return 0;

    int n = strsSize;
    int m = strlen(strs[0]);

    int resolved[n - 1];
    memset(resolved, 0, sizeof(resolved));

    int unresolved = n - 1;
    int deletions = 0;

    for (int col = 0; col < m && unresolved > 0; col++)
    {
        int needDelete = 0;

        for (int row = 0; row < n - 1; row++)
        {
            if (!resolved[row] && strs[row][col] > strs[row + 1][col])
            {
                needDelete = 1;
                break;
            }
        }

        if (needDelete)
        {
            deletions++;
            continue;
        }

        for (int row = 0; row < n - 1; row++)
        {
            if (!resolved[row] && strs[row][col] < strs[row + 1][col])
            {
                resolved[row] = 1;
                unresolved--;
            }
        }
    }

    return deletions;
}

File: problem/Delete_Columns_Sorted_II/minDeletionSize.py
class Solution(object):
    def minDeletionSize(self, strs):
        n = len(strs)
        m = len(strs[0])

        resolved = [False] * (n - 1)
        unresolved = n - 1
        deletions = 0

        for col in range(m):
            if unresolved == 0:
                break

            need_delete = False

            for row in range(n - 1):
                if not resolved[row] and strs[row][col] > strs[row + 1][col]:
                    need_delete = True
                    break

            if need_delete:
                deletions += 1
                continue

            for row in range(n - 1):
                if not resolved[row] and strs[row][col] < strs[row + 1][col]:
                    resolved[row] = True
                    unresolved -= 1

        return deletions